File: streams/examples/04-in-window-dedup.js
/**
 * Example 04 — In-window dedup.
 *
 * Source `webhooks.raw` may contain duplicate `eventId`s within a 5-minute
 * window. We dedupe per partition and emit a single record per unique
 * eventId per window onto `webhooks.unique`.
 *
 * Run:
 *   QUEEN_URL=http://localhost:6632 node examples/04-in-window-dedup.js
 */

import { Queen } from 'queen-mq'
import { Stream } from '@queenmq/streams'

const url = process.env.QUEEN_URL || 'http://localhost:6632'
const q = new Queen(url)

const handle = await Stream
  .from(q.queue('webhooks.raw'))
  .windowTumbling({ seconds: 300 })
  .reduce(
    (seen, m) => {
      const id = m.data && m.data.eventId
      if (!id) return seen
      if (!seen.includes(id)) seen.push(id)
      return seen
    },
    /* initial seen = */ []
  )
  .flatMap(seen => seen.map(eventId => ({ eventId })))
  .to(q.queue('webhooks.unique'))
  .run({
    queryId: 'examples.webhooks.dedup_5min',
    url
  })

await new Promise(r => setTimeout(r, 10 * 60_000))
console.log('Stats:', handle.metrics())
await handle.stop()
await q.close()

File: clients/client-py/examples/04_in_window_dedup.py
"""
Example 04 — In-window dedup. 1:1 port of 04-in-window-dedup.js.

Dedupes webhook eventIds within a 5-minute tumbling window per partition.

Run:
    QUEEN_URL=http://localhost:6632 python examples/04_in_window_dedup.py
"""

from __future__ import annotations

import asyncio
import os

from queen import Queen, Stream


URL = os.environ.get("QUEEN_URL", "http://localhost:6632")


def dedup(seen, m):
    eid = m.get("data", {}).get("eventId") if isinstance(m, dict) else None
    if not eid:
        return seen
    if eid not in seen:
        seen.append(eid)
    return seen


async def main():
    q = Queen(url=URL)
    try:
        handle = await (
            Stream.from_(q.queue("webhooks.raw"))
            .window_tumbling(seconds=300)
            .reduce(dedup, [])
            .flat_map(lambda seen, ctx=None: [{"eventId": eid} for eid in seen])
            .to(q.queue("webhooks.unique"))
            .run(query_id="examples.webhooks.dedup_5min", url=URL)
        )
        await asyncio.sleep(10 * 60)
        print("Stats:", handle.metrics())
        await handle.stop()
    finally:
        await q.close()


if __name__ == "__main__":
    asyncio.run(main())
